### PR TITLE
native: implement HF-based update

### LINKED ISSUE #3213: Support HF-based native contract update
Port https://github.com/neo-project/neo/pull/2942 and enable/refactor tests that were disabled in https://github.com/nspcc-dev/neo-go/pull/3212.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Let our natives be updated! Close #3213.

TODO:

- [x] Implement Oracle service cache update.
- [x] Make lazy MD cache not so lazy and remove generic Methods/Events.
- [x] Use cached HF-specific MD for new HF-specific MD initialization if no changes are required.
- [x] Enable Cockatrice with all dependent logic.
- [x] Tests (unit and state compatbility).

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@roman-khimov, the implementation seems to be in the final stage, I'm going to test it against C# node states for mainnet/testnet, and if everything is OK, then write a couple of unit tests.

File: pkg/core/native/management.go [hunk 1]
```diff
@@ -583,27 +596,80 @@ func updateContractCache(cache *ManagementCache, cs *state.Contract) {
...
+					isUpdate = true
+					activeIn = &hf // reuse ActiveIn variable for the initialization hardfork.
+					// Break immediately since native Initialize should be called only for the first hardfork in a raw
+					// (if there are multiple hardforks with the same enabling height).
+					// @roman-khimov, it looks like a bug in C# to me, but currently this behaviour doesn't affect us,
+					// because all natives perform some initialization-related actions only on ActiveIn hardfork trigger.
+					// But to me, it should be fixed. I'd suggest to call Initialize in cycle starting from activeIn and ending
+					// in latestHF. What do you think?
+					break
+				}
```
Comment (AnnaShaleva) [inline pkg/core/native/management.go:620]: Also need your opinion on this.

**Comment by AnnaShaleva** [comment]
With the latest C# node fixes mainnet states match perfectly (with Cockatrace enabled at 5210001 height):
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ go run scripts/compare-dumps/compare-dumps.go ./dumps/GoDump_mainnet_5224362/ ./dumps/Storage_334f454e_mainnet_5225043_cockatrice/
Processing directory BlockStorage_0
Processing directory BlockStorage_100000
Processing directory BlockStorage_200000
Processing directory BlockStorage_300000
Processing directory BlockStorage_400000
Processing directory BlockStorage_500000
Processing directory BlockStorage_600000
Processing directory BlockStorage_700000
Processing directory BlockStorage_800000
Processing directory BlockStorage_900000
Processing directory BlockStorage_1000000
Processing directory BlockStorage_1100000
Processing directory BlockStorage_1200000
Processing directory BlockStorage_1300000
Processing directory BlockStorage_1400000
Processing directory BlockStorage_1500000
Processing directory BlockStorage_1600000
Processing directory BlockStorage_1700000
Processing directory BlockStorage_1800000
Processing directory BlockStorage_1900000
Processing directory BlockStorage_2000000
Processing directory BlockStorage_2100000
Processing directory BlockStorage_2200000
Processing directory BlockStorage_2300000
Processing directory BlockStorage_2400000
Processing directory BlockStorage_2500000
Processing directory BlockStorage_2600000
Processing directory BlockStorage_2700000
Processing directory BlockStorage_2800000
Processing directory BlockStorage_2900000
Processing directory BlockStorage_3000000
Processing directory BlockStorage_3100000
Processing directory BlockStorage_3200000
Processing directory BlockStorage_3300000
Processing directory BlockStorage_3400000
Processing directory BlockStorage_3500000
Processing directory BlockStorage_3600000
Processing directory BlockStorage_3700000
Processing directory BlockStorage_3800000
Processing directory BlockStorage_3900000
Processing directory BlockStorage_4000000
Processing directory BlockStorage_4100000
Processing directory BlockStorage_4200000
Processing directory BlockStorage_4300000
Processing directory BlockStorage_4400000
Processing directory BlockStorage_4500000
Processing directory BlockStorage_4600000
Processing directory BlockStorage_4700000
Processing directory BlockStorage_4800000
Processing directory BlockStorage_4900000
Processing directory BlockStorage_5000000
Processing directory BlockStorage_5100000
Processing directory BlockStorage_5200000
Processing directory BlockStorage_5300000
file BlockStorage_5300000/dump-block-5225000.json: dump files differ in size: 362 vs 1000
compare-dumps dumpDirA dumpDirB
exit status 1
```

File: pkg/core/native/management.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/native/management.go]: Ref. https://github.com/neo-project/neo/issues/3200.

File: pkg/core/interop/context.go [hunk 2]
```diff
@@ -150,101 +157,218 @@ type MethodAndPrice struct {
...
-}
-
-// NewContractMD returns Contract with the specified list of methods.
-func NewContractMD(name string, id int32) *ContractMD {
+	// Events is a generic set of contract events with activation hardforks. Any HF-dependent part of events must not be
+	// used, there's a mdCache field for that.
+	Events []Event
+	// ActiveHFs is a map of hardforks that contract should react to. Contract update should be called for active
+	// hardforks. Note, that unlike the C# implementation, this map doesn't include contract's activation hardfork.
+	// This map is being initialized on contract creation and used as a read-only, hens, not protected
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:202]: hence

File: pkg/core/interop/context.go [hunk 3]
```diff
@@ -140,8 +141,14 @@ type Function struct {
 // Method is a signature for a native method.
 type Method = func(ic *Context, args []stackitem.Item) stackitem.Item
 
-// MethodAndPrice is a native-contract method descriptor.
+// MethodAndPrice is a generic hardfork-independent native contract method descriptor.
 type MethodAndPrice struct {
+	HFSpecificMethodAndPrice
+	ActiveFrom *config.Hardfork
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:146]: I'm not sure why are you keeping all of them this way. You can have the one and only proper contract version at the current height and then just switch it when HF happens.

Historic calls maybe? These do require some dances.

File: pkg/core/interop/context.go [hunk 4]
```diff
@@ -150,101 +157,218 @@ type MethodAndPrice struct {
...
+	}
+	c.mdCacheLock.RUnlock()
+
+	md := c.buildHFSpecificMD(hf)
+	return md
+}
+
+// buildHFSpecificMD builds hardfork-specific contract descriptor that includes methods and events active starting from
+// the specified hardfork or older.
+func (c *ContractMD) buildHFSpecificMD(hf *config.Hardfork) *HFSpecificContractMD {
```
Comment (roman-khimov) [inline pkg/core/interop/context.go]: Could it be `contract.MDAtHF()` with keeping HF-specific copies in contracts?

File: pkg/core/native/designate.go [hunk 5]
```diff
@@ -120,17 +119,23 @@ func newDesignate(p2pSigExtensionsEnabled bool, initialNodeRoles map[noderoles.R
...
 
 	return s
 }
 
 // Initialize initializes Designation contract. It is called once at native Management's OnPersist
 // at the genesis block, and we can't properly fill the cache at this point, as there are no roles
 // data in the storage.
-func (s *Designate) Initialize(ic *interop.Context) error {
+func (s *Designate) Initialize(ic *interop.Context, hf *config.Hardfork) error {
+	if hf != s.ActiveIn() {
```
Comment (roman-khimov) [inline pkg/core/native/designate.go:137]: Checking here...

File: pkg/core/native/crypto.go [hunk 6]
```diff
@@ -310,7 +309,7 @@ func (c *Crypto) Metadata() *interop.ContractMD {
 }
 
 // Initialize implements the Contract interface.
-func (c *Crypto) Initialize(ic *interop.Context) error {
+func (c *Crypto) Initialize(ic *interop.Context, hf *config.Hardfork) error {
```
Comment (roman-khimov) [inline pkg/core/native/crypto.go]: ...but not here.

File: pkg/config/hardfork.go [hunk 7]
```diff
@@ -5,6 +5,13 @@ package config
 // Hardfork represents the application hard-fork identifier.
 type Hardfork byte
 
+// HFDefault is a default value of Hardfork enum. It's a special constant
+// aimed to denote the node code enabled by default starting from the
+// genesis block. HFDefault is not a hard-fork, but this constant can be used for
+// convenient hard-forks comparison and to refer to the default hard-fork-less
+// node behaviour.
+const HFDefault Hardfork = 0
```
Comment (roman-khimov) [inline pkg/config/hardfork.go:13]: `// Default` for a better string representation.

File: pkg/core/interop/context.go [hunk 8]
```diff
@@ -204,10 +203,9 @@ type ContractMD struct {
 	// by mutex.
 	ActiveHFs map[config.Hardfork]struct{}
 
-	// mdCache contains hardfork-specific ready-to-use contract descriptors. This cache is lazy and thus, protected by
-	// mdCacheLock.
-	mdCacheLock sync.RWMutex
-	mdCache     map[config.Hardfork]*HFSpecificContractMD
+	// mdCache contains hardfork-specific ready-to-use contract descriptors. This cache is initialized in the native
+	// contracts constructors, and acts as read-only during the whole node lifetime, thus not protected by mutex.
+	mdCache map[config.Hardfork]*HFSpecificContractMD
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:208]: Yeah, this one is closer to what I was thinking about.

File: docs/node-configuration.md [hunk 9]
```diff
@@ -332,7 +332,7 @@ protocol-related settings described in the table below.
...
.../pull/3080) (ported from the [reference](https://github.com/neo-project/neo/pull/2883)) that increases `stackitem.Integer` JSON parsing precision up to the maximum value supported by the NeoVM. It also includes [#3085](https://github.com/nspcc-dev/neo-go/pull/3085) (ported from the [reference](https://github.com/neo-project/neo/pull/2810)) that enables strict check for notifications emitted by a contract to precisely match the events specified in the contract manifest. <br>• `Cockatrice` represents hard-fork introduced in [#3402](https://github.com/nspcc-dev/neo-go/pull/3402) (ported from the [reference](https://github.com/neo-project/neo/pull/2942)). Initially it is introduced along with the ability to update native contracts. This hard-fork also includes a couple of new native smart contract APIs: `keccaak256` of native CryptoLib contract introduced in [#3301](https://github.com/nspcc-dev/neo-go/pull/3301) (ported from the [reference](https://github.com/neo-project/neo/pull/2925)) and `getCommitteeAddress` of native NeoToken contract inctroduced in [#3362](https://github.com/nspcc-dev/neo-go/pull/3362) (ported from the [reference](https://github.com/neo-project/neo/pull/3154)). |
```
Comment (roman-khimov) [inline docs/node-configuration.md:335]: keccak

File: pkg/core/blockchain.go [hunk 10]
```diff
@@ -535,8 +536,10 @@ func (bc *Blockchain) init() error {
 			return fmt.Errorf("failed to check native %s state against autogenerated one: %w", md.Name, err)
 		}
 		if !bytes.Equal(storedCSBytes, autogenCSBytes) {
-			return fmt.Errorf("native %s: version mismatch (stored contract state differs from autogenerated one), "+
-				"try to resynchronize the node from the genesis", md.Name)
+			storedJ, _ := json.Marshal(storedCS)
+			autogenJ, _ := json.Marshal(autogenCS)
+			return fmt.Errorf("native %s: version mismatch for the latest hardfork %s (stored contract state differs from autogenerated one), "+
+				"try to resynchronize the node from the genesis: %s vs %s", md.Name, current, string(storedJ), string(autogenJ))
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:542]: I'm not sure it's a friendly error message. Many users may even not notice the real message in these JSON dumps.

File: pkg/core/native/nativehashes/hashes.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"github.com/nspcc-dev/neo-go/pkg/core/native/nativenames"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// Hashes of all native contracts.
+var (
+	Management  util.Uint160
```
Comment (roman-khimov) [inline pkg/core/native/nativehashes/hashes.go:11]: Just out of curiosity, can we have them precomputed automatically and saved as `const`s? `//go:generate` of some kind. We can combine them then with `nativenames` probably, consts don't cost a lot.

File: pkg/core/native/management.go [hunk 12]
```diff
@@ -653,14 +654,14 @@ func (m *Management) OnPersist(ic *interop.Context) error {
...
-		if err := native.Initialize(ic, activeIn); err != nil {
+		if err := native.Initialize(ic, activeIn, hfSpecificMD); err != nil {
 			return fmt.Errorf("initializing %s native contract at HF %d: %w", md.Name, activeIn, err)
 		}
 		if cache == nil {
 			cache = ic.DAO.GetRWCache(m.ID).(*ManagementCache)
 		}
 		updateContractCache(cache, cs)
-		
+
```
Comment (roman-khimov) [inline pkg/core/native/management.go:664]: Ouch.

**Review by roman-khimov — COMMENTED** [review]
Should be OK.

File: pkg/core/interop/context.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/core/interop/context.go:146]: > You can have the one and only proper contract version at the current height

That was an early version of this PR. And yes, due to the historic calls. To manage single contract version at a time, we need to initialize it every time when new historic call is processed (which I suppose is far from the best solution) or we need some kind of cache. So if we're to implement this cache anyway, then it's better to initialize it once during the node start in order not to affect the node performance during blocks processing. And, as discussed, we don't have a lot of natives/hardforks, so the cache size is not a problem for us (and we store pointers in cache anyway, as a result the most part of CS structures are shared between hardforks).

But that's just my thoughts. We can get back to `one and only proper contract version`.

File: pkg/core/interop/context.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/core/interop/context.go]: No, because we need methods offsets to be resent at this MD.

File: pkg/core/native/designate.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/core/native/designate.go:137]: What do you think?

File: pkg/core/blockchain.go (hunk 10 above)
Comment (AnnaShaleva) [reply pkg/core/blockchain.go:542]: These dumps are not indented, so they are quite compact, I've tested it. I may add some indent. Or is it better to remove dumps at all?

File: pkg/config/hardfork.go (hunk 7 above)
Comment (AnnaShaleva) [reply pkg/config/hardfork.go]: Fixed.

File: docs/node-configuration.md (hunk 9 above)
Comment (AnnaShaleva) [reply docs/node-configuration.md]: Fixed.

File: pkg/core/native/nativehashes/hashes.go (hunk 11 above)
Comment (AnnaShaleva) [reply pkg/core/native/nativehashes/hashes.go]: Check out the last commit, it's an interesting experiment, but I still want to keep `Uint160` variables to reuse them in other code.

> We can combine them then with nativenames probably,

They have different names, is it OK? E.g. `nativenames.Gas` vs `nativehashes.GasToken`.

File: pkg/core/native/nativehashes/gen.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+// package nativehashes contains hashes of all native contracts in their LE and Uint160 representation.
+package nativehashes
+
+import "github.com/nspcc-dev/neo-go/pkg/util"
+
+// Hashes of all native contracts in the LE string representation.
+const(
+{{- range .Natives }}
+	// {{ .Name }}StringLE is a hash of native {{ .Name }} contract in the LE string representation.
+	{{ .Name }}StringLE = {{ printf "%q" .HashLE }}
```
Comment (roman-khimov) [inline pkg/core/native/nativehashes/gen.go:30]: Do you need it?

File: pkg/core/native/nativehashes/gen.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+{{- end }}
+)
+
+// Hashes of all native contracts.
+var (
+{{- range .Natives }}
+	// {{ .Name }} is a hash of native {{ .Name }} contract.
+	{{ .Name }} util.Uint160
```
Comment (roman-khimov) [inline pkg/core/native/nativehashes/gen.go:38]: You can instantiate them as `util.Uint160{0xca, 0xfe...}`

File: pkg/core/native/nativehashes/gen.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+{{- end }}
+)
+
+func init() {
+{{- range .Natives }}
+	{{ .Name }}, _ = util.Uint160DecodeStringLE({{ .Name }}StringLE)
```
Comment (roman-khimov) [inline pkg/core/native/nativehashes/gen.go:44]: Without computations.

File: pkg/core/native/nativehashes/hashes.go [hunk 16]
```diff
@@ -1,34 +1,69 @@
+// Code generated by "go generate go run gen.go"; DO NOT EDIT.
```
Comment (roman-khimov) [inline pkg/core/native/nativehashes/hashes.go:1]: Let's push this patch into another PR.

File: pkg/core/native/nativehashes/gen.go (hunk 13 above)
Comment (AnnaShaleva) [reply pkg/core/native/nativehashes/gen.go:30]: I thought that maybe someone can make use of strings, but you're right, it seems to be extensive, we don't need it at all.

File: pkg/core/native/nativehashes/hashes.go (hunk 16 above)
Comment (AnnaShaleva) [reply pkg/core/native/nativehashes/hashes.go]: Ref. https://github.com/nspcc-dev/neo-go/pull/3431.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**